include/ctrace/ctrace.h

```diff
@@ -0,0 +1,283 @@
+#ifndef CTRACE_H
+#define CTRACE_H
+
+#include <limits.h>
+#include <setjmp.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#if defined(__cplusplus)
+# define C_LINKAGE extern "C"
+# define BEGIN_CTRACE C_LINKAGE {
+# define END_CTRACE }
+#else
+# define C_LINKAGE
+# define BEGIN_CTRACE
+# define END_CTRACE
+#endif
+
+#if !defined(__cplusplus)
+# if defined(__STDC_VERSION__)
+#  if __STDC_VERSION__ >= 201112L
+#   define CTRACE_C11_FEATURES ON
+#  else
+#   define CTRACE_C11_FEATURES OFF
+#  endif
+# else
+#  define CTRACE_C11_FEATURES OFF
+# endif
+#else
+# define CTRACE_C11_FEATURES OFF
+#endif
+
+#if defined(__cplusplus)
+# define CTRACE_NORETURN
+#elif CTRACE_C11_FEATURES == ON
+# include <stdnoreturn.h>
+# define CTRACE_NORETURN noreturn
+#else
+# define CTRACE_NORETURN
+#endif
+
+#if defined(_WIN32) || defined(__CYGWIN__)
+# define CTRACE_API C_LINKAGE __declspec(dllexport)
+#else
+# define CTRACE_API C_LINKAGE
+#endif
+
+#define ON 1
+#define OFF 0
+
+#if !defined(CTRACE_EXCEPTIONS)
+# define CTRACE_EXCEPTIONS OFF
+#endif
+
+#if !defined(CTRACE_FREESTANDING)
+# if defined(__STDC_HOSTED__)
+#  if __STDC_HOSTED__ == 1
+#   define CTRACE_FREESTANDING OFF
+#  else
+#   define CTRACE_FREESTANDING ON
+#  endif
+# else
+#  define CTRACE_FREESTANDING OFF
+# endif
+#endif
+
+#define CTRACE_BITCOUNT(x) (sizeof(x) * CHAR_BIT)
+#define EFLAG(n) 1 << n
+
+#define ESC     "\033["
+#define RESET   ESC "0m"
+#define RED     ESC "31m"
+#define GREEN   ESC "32m"
+#define YELLOW  ESC "33m"
+#define BLUE    ESC "34m"
+#define MAGENTA ESC "35m"
+#define CYAN    ESC "36m"
+
+BEGIN_CTRACE
+    typedef enum { cfalse, ctrue } ctrace_boolean;
+    typedef const char *ctrace_format_t;
+
+    // general::string:
+    typedef struct {
+        const void *data;
+        unsigned type;
+    } ctrace_string;
+
+    struct ctrace_buffered_string;
+    struct ctrace_exception_ctx;
+
+    /* Signifies `ctrace_free_string` should not be called.
+     */
+    typedef ctrace_string ctrace_managed_string;
```

Comment:
I have a concern about a custom string representation here. C strings are bad however they are simple, and it would be nice if the interface was simple for the library. I am thinking almost all of the time this stuff shouldn't be performance-sensitive. What do you think?